modules/stochastic_tools/src/utils/GaussianProcess.C

```diff
@@ -0,0 +1,371 @@
+
+  _covariance_function->buildHyperParamMap(_hyperparam_map, _hyperparam_vec_map);
+}
+
+void
+GaussianProcess::setupStoredMatrices(const RealEigenMatrix & input)
+{
+  _K_cho_decomp = _K.llt();
+  _K_results_solve = _K_cho_decomp.solve(input);
+}
+
+void
+GaussianProcess::generateTuningMap(const std::vector<std::string> & params_to_tune,
+                                   const std::vector<Real> & min_vector,
+                                   const std::vector<Real> & max_vector)
+{
+  _num_tunable = 0;
+
+  const bool upper_bounds_specified = min_vector.size();
+  const bool lower_bounds_specified = max_vector.size();
+
+  for (const auto param_i : index_range(params_to_tune))
+  {
+    const auto & hp = params_to_tune[param_i];
+    if (_covariance_function->isTunable(hp))
+    {
+      unsigned int size;
+      Real min;
+      Real max;
+      // Get size and default min/max
+      const bool found = _covariance_function->getTuningData(hp, size, min, max);
+
+      if (!found)
+        ::mooseError("The covariance parameter ", hp, " could not be found!");
+
+      // Check for overridden min/max
+      min = lower_bounds_specified ? min_vector[param_i] : min;
+      max = upper_bounds_specified ? max_vector[param_i] : max;
+      // Save data in tuple
+      _tuning_data[hp] = std::make_tuple(_num_tunable, size, min, max);
+      _num_tunable += size;
+    }
+  }
+}
+
+void
+GaussianProcess::standardizeParameters(RealEigenMatrix & data, bool keep_moments)
+{
+  if (!keep_moments)
+    _param_standardizer.computeSet(data);
+  _param_standardizer.getStandardized(data);
+}
+
+void
+GaussianProcess::standardizeData(RealEigenMatrix & data, bool keep_moments)
+{
+  if (!keep_moments)
+    _data_standardizer.computeSet(data);
+  _data_standardizer.getStandardized(data);
+}
+
+void
+GaussianProcess::tuneHyperParamsAdam(const RealEigenMatrix & training_params,
+                                     const RealEigenMatrix & training_data,
+                                     const GPOptimizerOptions & opts)
+{
+  std::vector<Real> theta(_num_tunable, 0.0);
+  _covariance_function->buildHyperParamMap(_hyperparam_map, _hyperparam_vec_map);
+
+  mapToVec(_tuning_data, _hyperparam_map, _hyperparam_vec_map, theta);
+
+  // Internal params for Adam; set to the recommended values in the paper
+  Real b1 = opts.b1;
+  Real b2 = opts.b2;
+  Real eps = opts.eps;
+
+  std::vector<Real> m0(_num_tunable, 0.0);
+  std::vector<Real> v0(_num_tunable, 0.0);
+
+  Real new_val;
+  Real m_hat;
+  Real v_hat;
+  Real store_loss = 0.0;
+  std::vector<Real> grad1;
+
+  // Initialize randomizer
+  std::vector<unsigned int> v_sequence(training_params.rows());
+  std::iota(std::begin(v_sequence), std::end(v_sequence), 0);
+  RealEigenMatrix inputs(_batch_size, training_params.cols());
+  RealEigenMatrix outputs(_batch_size, training_data.cols());
+  if (opts.show_optimization_details)
+    Moose::out << "OPTIMIZING GP HYPER-PARAMETERS USING Adam" << std::endl;
```

Comment:
Shouldn't we be using console to output things so that it preserves the prefix within multiapps? I'm sure we probably won't ever use this as a multiapp, but it would be good for consistency

Reply:
GaussianPRocess is a utility function which doesn't inherit from MooseObject. Also, it doesn't know about any other object so we don't have a way to grab the console. Unless it is global or static somewhere.